Add: changelog-file.md to the Docs-for-Contributors-and-Maintainers section. (#222)
pyOpenSci/python-package-guide: documentation/repository-files/changelog-file.md

# Changelog.md Guide

## Introduction

> The `changelog.md` document serves as a valuable resource for developers and users alike to track the evolution of a project over time. Understanding the structure and purpose of a changelog helps users and contributors stay informed about new features, bug fixes, and other changes introduced in each release.

Reviewer: I think we will want to name this file `CHANGELOG.md` to match our other meta files (`CONTRIBUTING.md` and `CODE_OF_CONDUCT.md`). Can you please update all referenced here? Also it looks like we got it wrong in package-structure-code/python-package-distribution-files-sdist-wheel.md, can you bring that change into this PR?

## What is changelog.md?

> `changelog.md` is a file commonly found in Python packages. Its primary purpose is to provide a chronological record of significant changes made to the project with each new release. This document helps users understand what has been added, fixed, or modified in the latest version of the software. In addition, it allows the contributors and users to understand the versioning syntax used for the package.

Reviewer: ```suggestion
The primary purpose of `CHANGELOG.md` is to provide a record of notable changes made to the project with each new release. This document helps users understand what has been added, fixed, modified, or removed with each version of the software. In addition, it allows the contributors and users to understand the versioning syntax used for the package.
```

This is a great description. Not trying to pick on it, just emphasizing that there is no real time component to a changelog, except that it happens with releases. And that it's not only for major work, but for any change that could be considered relevant or noteworthy for someone.

> In addition, it allows the contributors and users to understand the versioning syntax used for the package.

can you expand upon this? Typically projects follow one versioning scheme that follows some external rules. I don't see how a changelog would inform the reader about version string syntax.

Author: Good point. The intent of the last sentence was to emphasize that the changelog can inform the user of what version of the package to install based on its contents. I agree it has nothing to do with the syntax.

> [Keep a Changelog.md](https://keepachangelog.com/en/1.1.0/) is great, simple resource for understanding what a changelog is, how to create a good changelog, and includes some pointers of things to avoid in a changelog.

Reviewer: ```suggestion
[Keep a Changelog.md](https://keepachangelog.com/en/1.1.0/) is a great, simple resource for understanding what a changelog is and how to create a good changelog. It also includes examples of things to avoid.
```

> ```{admonition} Semantic Versioning

Reviewer: Versioning is a big topic, so much so that we have a guide for it! https://www.pyopensci.org/python-package-guide/package-structure-code/python-package-versions.html

We should link there rather than try and define semver twice in this guide IMO. But a good tip could be that the version string of a release often makes "chapters" in a changelog.

Author: Thanks for the callout @ucodery . @lwasser mentioned the same exact point in the feedback above and I just made sure to add a call-out rather than duplicating the effort.

> ```{admonition} Semantic Versioning

Author: Totally missed this page! Thanks for the pointer. Just added a callout in the admonition

- **Versioning**: Clear identification of each release version using semantic versioning or another versioning scheme adopted by the project.

- **Release Date**: The date when each version was released to the public, providing context for the timeline of changes.

- **Change Categories**: Organizing changes into categories such as "Added," "Changed," "Fixed," and "Removed" to facilitate navigation and understanding.

- **Description of Changes**: A concise description of the changes made in each category, including new features, enhancements, bug fixes, and deprecated functionality.

- **Links to Issues or Pull Requests**: References to relevant issue tracker items or pull requests associated with each change, enabling users to access more detailed information if needed.

- **Upgrade Instructions**: Guidance for users on how to upgrade to the latest version, including any breaking changes or migration steps they need to be aware of.

- **Contributor Recognition**: Acknowledgment of contributors who made significant contributions to the release, fostering a sense of community and appreciation for their efforts.

> ## What does it look like?

Author: This is great! I will add a section titled "How do maintainers use it?" right before the "What does it look like?" section. Also, your writeup is already very descriptive and I'm not sure there is more to add to it.

> ## What does it look like?

Reviewer 2: ok fantastic!! thank you!

> ```markdown

Reviewer: This is great, but I think real examples are sometimes easier to follow than ones entirely made up. We must have a project already that has a well-maintained changelog (@lwasser ?) that we can copy as an example? We could clip it for the code block and link to the repo for curious reader to dig deeper.

Author: I think this is a great idea. Also, it does seem like the projects will benefit from this guide on changelogs since after going through almost all the [projects on PyOpenSci](https://www.pyopensci.org/python-packages.html), there is an clear trend of not having a standard for changelogs. Some projects have them as rst files, some projects have them as .txt file and many don't have a changeling at all!

I just added the changelog of [Devicely](https://github.com/hpi-dhc/devicely/blob/main/CHANGELOG.md) as it is one of the few markdown versions that also follows semver and is based on [keep a changelog](https://www.google.com/url?sa=t&source=web&rct=j&opi=89978449&url=https://keepachangelog.com/en/1.1.0/&ved=2ahUKEwj3jIHjpLaFAxVqHDQIHcbaAxUQFnoECAcQAQ&usg=AOvVaw2VAelWNVwZ18HnfkpI_8pa). The format is not perfect however, since the with the 1.1.0, it seems the project authors stopped using the Added / Changed / Removed subheadings. 

Please advice if we'd like to revert back to the template example as opposed to the real example from Devicely if needed.

Reviewer 2: i agree with you both - great suggestion. let's use the real world example rather than a made up one!! and i love that we are also highlighting one of our packages by doing this!

# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased] - yyyy-mm-dd

Here we write upgrading notes for brands. It's a team effort to make them as
straightforward as possible.

### Added
- [PROJECTNAME-XXXX](http://tickets.projectname.com/browse/PROJECTNAME-XXXX)
  MINOR Ticket title goes here.
- [PROJECTNAME-YYYY](http://tickets.projectname.com/browse/PROJECTNAME-YYYY)
  PATCH Ticket title goes here.

### Changed

### Fixed

## [1.2.4] - 2017-03-15

Here we would have the update steps for 1.2.4 for people to follow.

### Added

### Changed

- [PROJECTNAME-ZZZZ](http://tickets.projectname.com/browse/PROJECTNAME-ZZZZ)
  PATCH Drupal.org is now used for composer.

### Fixed

> - [PROJECTNAME-TTTT](http://tickets.projectname.com/browse/PROJECTNAME-TTTT)

Author: See [comment](https://github.com/pyOpenSci/python-package-guide/pull/222#discussion_r1558449643)

- **Contributor Recognition**: Acknowledgment of contributors who made significant contributions to the release, fostering a sense of community and appreciation for their efforts.

## How do maintainers use it?

Often you will see a change log that documents a few things

Unreleased commits at the top of the changelog, it's common to have an `## Unreleased` section. This is where you can add new fixes, updates and features that have been added to the package since the last release.

This section might look something like this:

> ```markdown

Author: This section was rewritten. I think we should follow the progress in this thread instead https://github.com/pyOpenSci/python-package-guide/pull/222#discussion_r1559801388

## Unreleased
* Fix: Fixed a bug.... more here. (@github_username, #issuenumber)
* Add: new feature to ...more here (@github_username, #issuenumber)

> Below the unreleased section, there will be a section for each release. In these sections, you can organize new Fixes, Additions and update together.

Author: I agree. I moved explanation out and cleaned up the section.

## v1.0

### Updates
* Fix: Fixed a bug.... more here. (@github_username, #issuenumber)

### Additions
* Add: new feature to ...more here (@github_username, #issuenumber)

### Deprecations

### Contributors to this release
```

> When you are ready to make a new release, you can re-organize the list of unreleased elements into a section that is specific to that new release number.

Reviewer 2: ahhhh @ucodery this might be partially my fault. 

[this is what i'm used to seeing](https://github.com/pyOpenSci/pyosMeta/blob/main/CHANGELOG.md)

or [this](https://github.com/stravalib/stravalib/blob/main/changelog.md)
 
essentially you have a large flowing unreleased section. Then when you go to make a release, you organize into fixed, added, removed/ deprecated contributors. 

However your suggest is also good and much simpler. i'm curious what you think? we could stick with the simpler version but add a small breakout on how the changelog might be reorganized for a release (as an option). 

Any changelog is better than no changelog so i'm open to thoughts here :)

> When you are ready to make a new release, you can re-organize the list of unreleased elements into a section that is specific to that new release number.

Author: Both valid points! I think it's easier to just see an example of where my thoughts are on this rather than explaining it. I rewrote the section and took somewhat of a hybrid approach. Let me know what you think!

> When you are ready to make a new release, you can re-organize the list of unreleased elements into a section that is specific to that new release number.

Reviewer: Okay, that makes more sense, thanks for the examples @lwasser!

> The unreleased section then always lives at the top of the file. And as new features are added after release 1.x, they get added to that section.

Reviewer: ```suggestion
The unreleased section then always lives at the top of the file and new features continue to be added there. At the same time, after releasing a version like 1.0 all of its features remain in that specific section.
```

## What does it look like?

> This example comes from [Devicely](https://github.com/hpi-dhc/devicely/tree/main), one of the PyOpenSci reviewed and accepted packages.

Reviewer: ```suggestion
This example comes from [Devicely](https://github.com/hpi-dhc/devicely/blob/main/CHANGELOG.md), a pyOpenSci accepted package.
```
